### PR TITLE
[Flow] Loosen restrictions for dequantization fusion

### LINKED ISSUE #15826: [DT][CPU] Some casting ops are not fused with mmt4d kernels
We set encodings on CastOpInterface ops, so we can fuse in cast ops to mmt4d dispatch. However, we only fuse cast ops when they are not "group dequantization" ops. It does not happen when it is something like `arith.truncf %lhs f32 to f16` + `f16.f16.f16` matmul. This results in additional dispatches. We are having `pack` dispatch, `arith.truncf` dispatch and `mmt4d` dispatch. We should either fix it in set_encoding or dispatch formation. In the former solution, we will have `arith.truncf + pack` dispatch and mmt4d dispatch. In the latter solution, we will have `[optional consumers] + pack` dispatch and `arith.truncf + mmt4d` dispatch.

To repro:

Run `iree-compile --output-format=vm-bytecode --iree-hal-target-backends=llvm-cpu --iree-llvmcpu-target-cpu=cascadelake --iree-llvmcpu-target-triple=x86_64-unknown-linux-gnu ~/repro.mlir -o /tmp/a.vmfb`

```mlir
#map = affine_map<(d0, d1) -> (d0, d1)>
module {
  func.func @main(%arg0: tensor<?x?xf32>, %arg1: tensor<?x?xf16>) -> tensor<?x?xf16> {
    %cst = arith.constant 0.000000e+00 : f16
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %dim = tensor.dim %arg0, %c0 : tensor<?x?xf32>
    %dim_0 = tensor.dim %arg0, %c1 : tensor<?x?xf32>
    %0 = tensor.empty(%dim, %dim_0) : tensor<?x?xf16>
    %1 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel"]} ins(%arg0 : tensor<?x?xf32>) outs(%0 : tensor<?x?xf16>) {
    ^bb0(%in: f32, %out: f16):
      %5 = arith.truncf %in : f32 to f16
      linalg.yield %5 : f16
    } -> tensor<?x?xf16>
    %dim_1 = tensor.dim %arg1, %c1 : tensor<?x?xf16>
    %2 = tensor.empty(%dim, %dim_1) : tensor<?x?xf16>
    %3 = linalg.fill ins(%cst : f16) outs(%2 : tensor<?x?xf16>) -> tensor<?x?xf16>
    %4 = linalg.matmul ins(%1, %arg1 : tensor<?x?xf16>, tensor<?x?xf16>) outs(%3 : tensor<?x?xf16>) -> tensor<?x?xf16>
    return %4 : tensor<?x?xf16>
  }
}
```

### PR DESCRIPTION
**Pull request by Max191**
The conditions for cloning dequantization ops into a dispatch
region are too conservative, and miss fusions of some
dequantization-like ops. This loosens the restrictions for
cloning dequantization-like ops into dispatch regions.

Closes https://github.com/openxla/iree/issues/15826

### CODE REVIEW

**Comment by Max191** [comment]
This is based on https://github.com/openxla/iree/pull/15662, and can wait until the PR has landed.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp
Comment (Max191) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:507]: This is the basic change from this PR. The previous dequantization matching has been replaced with this matcher.

**Comment by hanhanW** [comment]
Should we fix the first sub-task in https://github.com/openxla/iree/issues/15826#issuecomment-1846106427 first? So we can make sure that SetEncoding is not set on truncation ops?

**Comment by Max191** [comment]
> Should we fix the first sub-task in [#15826 (comment)](https://github.com/openxla/iree/issues/15826#issuecomment-1846106427) first? So we can make sure that SetEncoding is not set on truncation ops?

Yep, I can do that first! I just wanted to rebase this, since it was already written (although I still need to add tests). I'll fix SetEncoding and then I'll add tests here.

**Comment by hanhanW** [comment]
@Max191 is it rebased on https://github.com/openxla/iree/commit/ec853149c6a5f388d1ac217d5d58e7bff93310f9? If not, can you rebase on ToT so we can include the change?

**Comment by Max191** [comment]
> @Max191 is it rebased on [ec85314](https://github.com/openxla/iree/commit/ec853149c6a5f388d1ac217d5d58e7bff93310f9)? If not, can you rebase on ToT so we can include the change?

Ah, whoops! I forgot to rebase after I merged that. Thanks for reminding me.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:507]: This is not a trivial function, can you add a doc to it?

Comment (Max191) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:507]: I have some comments in RegionOpUtils.h. Should I add something here as well?

Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp:507]: oops, I missed that part. We don't need the comment in `.cpp`, thank you!

**Review by hanhanW — APPROVED** [review]

**Pull request merged**
